internal/DebounceDecorator/DebounceDecorator.js

```diff
@@ -60,66 +60,66 @@ const defaultConfig = {
 const DebounceDecorator = hoc(defaultConfig, (config, Wrapped) => {
 	const {cancel, debounce, delay} = config;
 
-	return class extends Component {
-		static displayName = 'DebounceDecorator';
-
-		static propTypes = /** @lends sandstone/internal/DebounceDecorator.DebounceDecorator.prototype */ {
-			/**
-			 * Handler for `onChange` events
-			 *
-			 * `'onChange'` can be changed to a different prop name by specifying the `debounce`
-			 * config option.
-			 *
-			 * @see {@link sandstone/internal/DebounceDecorator.DebounceDecorator.defaultConfig#debounce}
-			 * @name onChange
-			 * @memberof sandstone/internal/DebounceDecorator.DebounceDecorator.prototype
-			 * @type {Function}
-			 * @public
-			 */
-			[debounce]: PropTypes.func
-		};
-
-		constructor (props) {
-			super(props);
-			this.job = new Job(this.emitEvent.bind(this), delay);
-		}
-
-		componentWillUnmount () {
-			this.job.stop();
-		}
+	const Debounce = (props) => {
+		let debounceProps = props;
 
-		emitEvent (ev) {
-			if (this.props[debounce]) {
-				this.props[debounce](ev);
+		const emitEvent = useCallback((ev) => {
+			if (props[debounce]) {
+				props[debounce](ev);
 			}
-		}
-
-		handleEvent = (ev) => {
-			this.job.start(ev);
-		};
+		}, [props]);
 
-		handleCancel = (ev) => {
-			if (this.props[cancel]) {
-				this.props[cancel](ev);
-			}
-			this.job.stop();
-		};
+		const job = useMemo(() => new Job(emitEvent, delay), [emitEvent]);
```

Comment:
I have a deep concern that what would be happened when `props` is changed in any reason during debounce logic works (=== meaning job.start() is called and job.stop() has not been called). In this case, `emitEvent` is re-created and `job` object is also re-created so that the old `job` is garbage collected without calling stop() function. It could make a call to an old `emitEvent` function object that is saved in a closure when `delay` is passed. What do you think about this?

Reply:
Also, React does NOT guarentee that a cached object via `useMemo` is always kept when dependencies are not change. React will do it's best to keep the cached one but it does not guarantee. So, the returning value within `useMemo` should be pure: always same. But a job object returned from `new Job()` is not a pure object actually. It has a state by start and stop. We may need to find other safe option. Or, we can leave this code after we confirm that there is no risky case.